### PR TITLE
fix(Scripts/ZulGurub): Renataki - improvements:

### LINKED ISSUE #11620: [ZG] Edge of Madness - Renataki - Missing/Wrong spells
### Current Behaviour

Uses Arcane Devastation (ID: 34799)
Uses the wrong Vanish spell (ID: 41476)
Has the wrong Gouge (ID: 38863)
Has the wrong Ambush (ID: 34799)
Does not use Thousand Blades from what I've seen in testing (24649 and 24767, probably the latter)
Does not use Thrash (ID: 3391)
Does not Enrage when reaching 30% health.

### Expected Blizzlike Behaviour

Should not have Arcane Devastation (ID: 34799)
Should have Vanish (ID: 24699)
Should have Ambush (ID: 24337)
Should use an AoE Gouge and then proceed to ignore the gouged players (ID: 24698)
Should occasionally use Thousand Blades (24649 and 24767, probably the latter) 2 to 5 times (random amount) consecutively on random ranged players
Should occasionally use Thrash (ID: 3391)
Should Enrage after reaching 30% health with the emote "Renataki becomes enraged!"

### Source

Renataki has been the hardest so far because of a lack of sources and guides for his encounter.
Mostly used:
https://www.youtube.com/watch?v=urX62TDf1gM
https://www.youtube.com/watch?v=9lwDAPmswu0
https://www.youtube.com/watch?v=XgwrVqh76Us&
https://youtu.be/EKoS0GIuZtI

Thousand Blades being used 5 times in a row: https://youtu.be/urX62TDf1gM?t=125
ditto: https://youtu.be/XgwrVqh76Us?t=25

Classic:
https://www.youtube.com/watch?v=lQl2aV80pU0
https://www.youtube.com/watch?v=l_n7OgBVjWA

### Steps to reproduce the problem

If the Summoning Event is implemented:
`.event stop 27`
`.event stop 28`
`.event stop 30`
`.event start 29`
`.additem 19931`
`.go xyz -11905 -1911 66 309`
use the brazier and engage Renataki

else:
`.go xyz -11905 -1911 66 309`
`.npc add temp 15084`

### Extra Notes

Classic WoW has the Vanish mechanic breaking instantly.

I looked for the ID of Enrage but couldn't find it, unfortunately. Looking at Classic videos it *looks* like it's at least 50% Attack Speed Increase. Maybe ~~24318~~? 24318 is Mandokir's frenzy.

Edit: From reviewing the sources I've noticed that the timers *look* like:
Vanish: 25s for the first, 38-45s for the following ones.
Thousand Blades: 15-21s
Gouge: 5-10s for the first one. 11-15s or 10-16s for the following ones,  average of 13s

As for Thrash, I did not find a good video showcasing it, sorry.

### AC rev. hash/commit

https://github.com/azerothcore/azerothcore-wotlk/commit/d1b690e3fa89897b6eada5fc8ec11354227cf2d5

### Operating system

Windows 10

### Custom changes or Modules

_No response_

### LINKED ISSUE #11621: [ZG] Edge of Madness - Renataki - Vanish & Ambush mechanic
### Current Behaviour

The wrong Vanish is used and boss does not have Ambush.

### Expected Blizzlike Behaviour

Renataki should use vanish 25s after starting combat and every 38-45s following that, moving to a random player and using Ambush to take out a major portion of their health.
Pets continue to attack even while he's invisible.

### Source

Pets attacking while Vanished: https://youtu.be/urX62TDf1gM?t=153
AoE does not immediately cancel Vanish: https://youtu.be/9lwDAPmswu0?t=26
Successful Vanish + Ambush in WoW Classic: https://youtu.be/lQl2aV80pU0?t=38 

https://www.youtube.com/watch?v=urX62TDf1gM has Vanish used at 0:30, 1:10, 1:55, 2:35, 3:13
https://www.youtube.com/watch?v=9lwDAPmswu0
https://www.youtube.com/watch?v=XgwrVqh76Us&
https://youtu.be/EKoS0GIuZtI

### Steps to reproduce the problem

If the Summoning Event is implemented:
`.event stop 27`
`.event stop 28`
`.event stop 30`
`.event start 29`
`.additem 19931`
`.go xyz -11905 -1911 66 309`
use the brazier and engage Renataki

else:
`.go xyz -11905 -1911 66 309`
`.npc add temp 15084`

### Extra Notes

Vanish (ID: 24699)
Ambush (ID: 24337)

WoW Classic seems to be bugged with the boss immediately cancelling Vanish from the slightest amount of damage, making this mechanic unseen in Classic videos.

Wowpedia and Wowhead claims that AoE *can* break Vanish, but I don't know if it's a chance on hit or a damage threshold.

### AC rev. hash/commit

https://github.com/azerothcore/azerothcore-wotlk/commit/d1b690e3fa89897b6eada5fc8ec11354227cf2d5

### Operating system

Windows 10

### Custom changes or Modules

_No response_

### PR DESCRIPTION
**Pull request by UltraNix**
Added missing spells and events.
Removed invalid spells
Corrected event timers
Fixes #11620
Fixes #11621

## Issues Addressed:

- Closes #11620
- Closes #11621

## Tests Performed:

- Not tested.

## How to Test the Changes:

If the Summoning Event is implemented:
`.event stop 27`
`.event stop 28`
`.event stop 30`
`.event start 29`
`.additem 19931`
`.go xyz -11905 -1911 66 309`
use the brazier and engage Renataki

else:
`.go xyz -11905 -1911 66 309`
`.npc add temp 15084`

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

**Comment by Gultask** [comment]
- [ ] If everyone is hit by Gouge, the encounter ends and the boss resets.
 - [ ] Thousand Blades is only being cast once, and only on the main target from what I saw. It should be cast several times back-to-back, and changing targets if possible (3 to 5 times or maybe a constant 4 times to be more consistent) I also think it should prefer ranged targets whenever possible but I'm not 100% sure about that.
 - [ ] Vanish doesn't stop the boss from attacking the tank so stealth is broken immediately. I think he should stop doing normal melee attacks completely while stealthed, only breaking stealth for Ambush.

As for more minor issues:

 - [ ] Pets should be able to attack the boss while he's stealthed, currently they leave him alone, returning to their master instead.

**Comment by UltraNix** [comment]
@Gultask Please, test it.

**Comment by Gultask** [comment]
I got a build error:

```
3>I:\AC\src\common\Utilities\Containers.h(100,21): error C2678: binary '=': no operator found which takes a left-hand operand of type 'const ObjectGuid' (or there is no acceptable conversion) (compiling source file I:\AC\src\server\scripts\EasternKingdoms\ZulGurub\boss_renataki.cpp)
3>I:\AC\src\server\game\Entities\Object\ObjectGuid.h(252,1): message : could be 'ObjectGuid &ObjectGuid::operator =(ObjectGuid &&)' (compiling source file I:\AC\src\server\scripts\EasternKingdoms\ZulGurub\boss_renataki.cpp)
3>I:\AC\src\server\game\Entities\Object\ObjectGuid.h(252,1): message : or       'ObjectGuid &ObjectGuid::operator =(const ObjectGuid &)' (compiling source file I:\AC\src\server\scripts\EasternKingdoms\ZulGurub\boss_renataki.cpp)
3>I:\AC\src\common\Utilities\Containers.h(82,1): message : while trying to match the argument list '(const ObjectGuid, const ObjectGuid)' (compiling source file I:\AC\src\server\scripts\EasternKingdoms\ZulGurub\boss_renataki.cpp)
3>I:\AC\src\server\scripts\EasternKingdoms\ZulGurub\boss_renataki.cpp(190): message : see reference to function template instantiation 'void Acore::Containers::RandomResize<GuidSet>(C &,size_t)' being compiled
3>        with
3>        [
3>            C=GuidSet
3>        ]
3>Done building project "scripts.vcxproj" -- FAILED.

```


Untitled

**Comment by UltraNix** [comment]
@Gultask Try now.

**Comment by Gultask** [comment]
```
3>I:\AC\src\server\scripts\EasternKingdoms\ZulGurub\boss_renataki.cpp(158,1): error C2664: 'void std::list<Unit *,std::allocator<Unit *>>::push_back(_Ty &&)': cannot convert argument 1 from 'ObjectGuid' to '_Ty &&'
3>        with
3>        [
3>            _Ty=Unit *
3>        ]
3>I:\AC\src\server\scripts\EasternKingdoms\ZulGurub\boss_renataki.cpp(158,1): message : Reason: cannot convert from 'ObjectGuid' to '_Ty'
3>        with
3>        [
3>            _Ty=Unit *
3>        ]
3>I:\AC\src\server\scripts\EasternKingdoms\ZulGurub\boss_renataki.cpp(158,77): message : No user-defined-conversion operator available that can perform this conversion, or the operator cannot be called
3>C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.32.31326\include\list(957,10): message : see declaration of 'std::list<Unit *,std::allocator<Unit *>>::push_back' (compiling source file I:\AC\src\server\scripts\EasternKingdoms\ZulGurub\boss_renataki.cpp)
4>authserver.vcxproj -> I:\Build\bin\RelWithDebInfo\authserver.exe
3>Done building project "scripts.vcxproj" -- FAILED.

```

**Comment by UltraNix** [comment]
Once again

**Comment by Gultask** [comment]
```
4>   Creating library I:/Build/src/server/apps/RelWithDebInfo/authserver.lib and object I:/Build/src/server/apps/RelWithDebInfo/authserver.exp
3>boss_renataki.cpp
3>I:\AC\src\server\scripts\EasternKingdoms\ZulGurub\boss_renataki.cpp(158,1): error C2664: 'void std::list<Unit *,std::allocator<Unit *>>::push_back(_Ty &&)': cannot convert argument 1 from 'ObjectGuid' to '_Ty &&'
3>        with
3>        [
3>            _Ty=Unit *
3>        ]
3>I:\AC\src\server\scripts\EasternKingdoms\ZulGurub\boss_renataki.cpp(158,1): message : Reason: cannot convert from 'ObjectGuid' to '_Ty'
3>        with
3>        [
3>            _Ty=Unit *
3>        ]
3>I:\AC\src\server\scripts\EasternKingdoms\ZulGurub\boss_renataki.cpp(158,77): message : No user-defined-conversion operator available that can perform this conversion, or the operator cannot be called
3>C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.32.31326\include\list(957,10): message : see declaration of 'std::list<Unit *,std::allocator<Unit *>>::push_back'
3>Done building project "scripts.vcxproj" -- FAILED.

```

**Comment by UltraNix** [comment]
WTF

**Comment by UltraNix** [comment]
Try again pull soure or rebase your local.

**Comment by Gultask** [comment]
I tried it again, creating another branch. 27 errors. Uh oh. 😬 

@temperrr  could you test it in the meantime? Or maybe Maelthyr if he's available?

**Comment by acidmanifesto** [comment]
> I tried it again, creating another branch. 27 errors. Uh oh. 😬
> 
> @temperrr could you test it in the meantime? Or maybe Maelthyr if he's available?

I'm doing a absolute clean setup and recompile right now. Will report back if i get any of the compile errors.

**Comment by acidmanifesto** [comment]
so after compiling i get this:
```
21>D:\New folder\azerothcore-wotlk-custom\src\common\Utilities\Containers.h(100,21): error C3892: 'keepIt': you cannot assign to a variable that is const (compiling source file D:\New folder\azerothcore-wotlk-custom\src\server\scripts\EasternKingdoms\ZulGurub\boss_renataki.cpp)
21>D:\New folder\azerothcore-wotlk-custom\src\server\scripts\EasternKingdoms\ZulGurub\boss_renataki.cpp(190): message : see reference to function template instantiation 'void Acore::Containers::RandomResize<std::set<Unit *,std::less<Unit *>,std::allocator<Unit *>>>(C &,size_t)' being compiled
21>        with
21>        [
21>            C=std::set<Unit *,std::less<Unit *>,std::allocator<Unit *>>
21>        ]
```
image

Verdict, Can not compile due to errors.
Windows 11, VS 2022

File: src/server/scripts/EasternKingdoms/ZulGurub/boss_renataki.cpp [hunk 1]
```diff
@@ -46,123 +53,180 @@ class boss_renataki : public CreatureScript
...
 
-                me->SetUnitFlag(UNIT_FLAG_NOT_SELECTABLE);
-                Invisible = true;
+            return true;
+        }
 
-                Invisible_Timer = urand(15000, 30000);
-            }
-            else Invisible_Timer -= diff;
+        bool CanBeSeen(Player const* player) override
```
Comment (acidmanifesto) [inline src/server/scripts/EasternKingdoms/ZulGurub/boss_renataki.cpp:94]: bool CanBeSeen(Player const* /*player*/) override

**Review by acidmanifesto — CHANGES_REQUESTED** [review]

File: src/server/scripts/EasternKingdoms/ZulGurub/boss_renataki.cpp (hunk 1 above)
Comment (acidmanifesto) [reply src/server/scripts/EasternKingdoms/ZulGurub/boss_renataki.cpp:94]: this is in regards to the ci testing ^^

**Comment by UltraNix** [comment]
@acidmanifesto Fixed.

**Comment by acidmanifesto** [comment]
@Gultask i will be unavailable for a the next day and a half, if you can test and let me know. Ultranix resolved the compile issues .

**Comment by temperrr** [comment]
[ ] Never vanishes (never uses ambush)
[ ] never uses gouge, or I think he tries to do it but it does not go on the chars (no combat log entries) but he does switch targets very briefly 
[x] uses thousand blades correctly

**Comment by temperrr** [comment]
Seems to work as described.

**Comment by acidmanifesto** [comment]
> Seems to work as described.

So its tested?

**Comment by UltraNix** [comment]
@temperrr Please test pets again.

**Comment by Maelthyrr** [comment]
> @temperrr Please test pets again.

Work ;)

**Review by acidmanifesto — APPROVED** [review]

**Comment by acidmanifesto** [comment]
Approved

File: src/server/scripts/EasternKingdoms/ZulGurub/boss_renataki.cpp [hunk 2]
```diff
@@ -46,123 +53,190 @@ class boss_renataki : public CreatureScript
...
         {
-            if (!UpdateVictim())
-                return;
-
-            //Invisible_Timer
-            if (Invisible_Timer <= diff)
+            if (!_enraged && HealthBelowPct(30))
             {
-                me->InterruptSpell(CURRENT_GENERIC_SPELL);
+                me->TextEmote("%s becomes enraged", me, false);
```
Comment (Nefertumm) [inline src/server/scripts/EasternKingdoms/ZulGurub/boss_renataki.cpp:80]: move this to db and then we are safe to merge

**Pull request merged**